modin/experimental/xgboost/xgboost_ray.py

```diff
@@ -0,0 +1,339 @@
+# Licensed to Modin Development Team under one or more contributor license agreements.
+# See the NOTICE file distributed with this work for additional information regarding
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+
+import time
+import logging
+from threading import Thread
+from typing import Dict, Optional
+from multiprocessing import cpu_count
+
+import xgboost as xgb
+import ray
+from ray.services import get_node_ip_address
+import numpy as np
+import pandas
+
+from modin.api import unwrap_partitions
+
+LOGGER = logging.getLogger("[modin.xgboost]")
+
+
+def _start_rabit_tracker(num_workers: int):
+    """Start Rabit tracker. The workers connect to this tracker to share
+    their results."""
+    host = get_node_ip_address()
+
+    env = {"DMLC_NUM_WORKER": num_workers}
+    rabit_tracker = xgb.RabitTracker(hostIP=host, nslave=num_workers)
+
+    # Get tracker Host + IP
+    env.update(rabit_tracker.slave_envs())
+    rabit_tracker.start(num_workers)
+
+    # Wait until context completion
+    thread = Thread(target=rabit_tracker.join)
+    thread.daemon = True
+    thread.start()
+
+    return env
+
+
+class RabitContext:
+    """Context to connect a worker to a rabit tracker"""
+
+    def __init__(self, actor_ip, args):
+        self.args = args
+        self.args.append(("DMLC_TASK_ID=[modin.xgboost]: " + actor_ip).encode())
+
+    def __enter__(self):
+        xgb.rabit.init(self.args)
+        LOGGER.info("-------------- rabit started ------------------")
+
+    def __exit__(self, *args):
+        xgb.rabit.finalize()
+        LOGGER.info("-------------- rabit finished ------------------")
+
+
+@ray.remote
+class ModinXGBoostActor:
+    def __init__(self, ip, nthread=cpu_count()):
+        self._evals = []
+        self._dpredict = []
+        self._ip = ip
+        self._nthreads = nthread
+
+        LOGGER.info(f"Actor <{self._ip}>, nthread = {self._nthreads} was initialized.")
+
+    def get_actor_ip(self):
+        return self._ip
+
+    def _get_dmatrix(self, *X_y):
+        s = time.time()
+        X = X_y[: len(X_y) // 2]
+        y = X_y[len(X_y) // 2 :]
+
+        assert (
+            len(X) == len(y) and len(X) > 0
+        ), "X and y should have the equal length more than 0"
+
+        X = pandas.concat(X, axis=0)
+        y = pandas.concat(y, axis=0)
```

Comment:
that's a lot of memory copying (because `X_y` is a tuple due to being an asterisk argument), can this be reduced somehow?

Reply:
Do you mean something like that?
```python
def _get_dmatrix(self, X_y):
    X = X_y[: len(X_y) // 2]
    y = X_y[len(X_y) // 2 :]
    ...
    
    
def set_train_data(self, *X_y, add_as_eval_method=None):
    data = list(X_y)
    self._dtrain = self._get_dmatrix(X_y)
```

Reply:
`*X_y`(tuple) was changed to `X_y`(list) by a way from comment above.

Reply:
I don't think it changes much, might be even changing it to being slower... some profiling and considerations of what is the input are needed. If `X_y` would be a numpy array then selecting half of it would be just making a view, which is almost zero-cost.

Reply:
Time measurements for approaches 'with asterisk/without asterisk' is the next:
```
Time of asterisk = 0.00313
Time of tuple is = 0.00162
```

<details><summary>Script for measurements</summary>

```python
import time
import numpy as np
from numpy.random import randn
import pandas

data = tuple([pandas.DataFrame(randn(10000)) for i in range(0, 100000)])

def func_asterisk_tuple(dat):
    def func(*tup):
        l1 = tup[: len(tup) // 2]
        l2 = tup[len(tup) // 2 :]

    func(*dat)


def func_tuple(dat):
    def func(tup):
        l1 = tup[: len(tup) // 2]
        l2 = tup[len(tup) // 2 :]

    func(dat)


t_asterisk = 0
t_tuple = 0
N = 10
for it in range(0, N):
    s = time.time()
    func_asterisk_tuple(data)
    t_asterisk += time.time() - s

    s = time.time()
    func_tuple(data)
    t_tuple += time.time() - s


print(f"Time of asterisk is {t_asterisk / N}")
print(f"Time of tuple is {t_tuple / N}")

```
</details>

Reply:
Cool, so "tuple" is approx. two times faster than asterisk.

If the data passed is a tuple of partitions then I think the overhead would be negligible, so we should go with the easiest implementation (I was worried the data would be raw items - this would mean tuples of hundreds of MB or more, here I assume it would be a few MB at most - no one would notice).